Safemode - Introduce TimeWindowCompactionStrategy Guardrails

fee-mendes opened the pull request:
This series introduces two configurable options when working with TWCS tables: 

- `restrict_twcs_default_ttl` - a LiveUpdate-able tri_mode_restriction which defaults to WARN and will notify the user whenever a TWCS table is created without a `default_time_to_live` setting
- `twcs_max_window_count` - Which forbids the user from creating TWCS tables whose window count (buckets) are past a certain threshold. We default to 50, which should be enough for most use cases, and a setting of 0 effectively disables the check.

Refs: #6923
Fixes: #9029

tzach commented:
> restrict_twcs_default_ttl - a LiveUpdate-able tri_mode_restriction which defaults to WARN and will notify the user whenever a TWCS table is created without a default_time_to_live setting

Is there a reason to allow TWCS without default_time_to_live?
I have two worries:
- warnings are ineffective
- one more configuration parameter

fee-mendes commented:
> Is there a reason to allow TWCS without default_time_to_live? I have two worries:

Yes, we can't be sure if the user will use `USING TTL` when ingesting data (although I agree at least a `default_time_to_live` should be explicitly set -- for a number of reason, but I don't think enforcing it should be the ideal). Plus, we don't want to block `ALTER TABLE` statements for tables that have been incorrectly created initially.

avikivity commented on cql3/statements/create_table_statement.cc:
```diff
@@ -452,6 +453,20 @@ std::optional<sstring> check_restricted_table_properties(
             break;
         }
     }
+    if (strategy && *strategy == sstables::compaction_strategy_type::time_window){
```
Our coding style wants a space between `)` and `{`.

This should be in the strategy itself, though I see you followed existing practice here.

avikivity commented on cql3/statements/create_table_statement.cc:
```diff
@@ -452,6 +453,20 @@ std::optional<sstring> check_restricted_table_properties(
+        int max_windows = qp.db().get_config().twcs_max_window_count();
+
+        if (ttl > 0){
+           int window_count = (int) ttl / _twcs_options.get_sstable_window_size().count();
```
We avoid C-style cases, they are dangerous. In this case, why cast at all? It just provides opportunity for overflow.

Time is best divided as two chrono::duration objects so we don't lose track of the units.

avikivity commented on cql3/statements/create_table_statement.cc:
```diff
@@ -452,6 +453,20 @@ std::optional<sstring> check_restricted_table_properties(
+        if (ttl > 0){
+           int window_count = (int) ttl / _twcs_options.get_sstable_window_size().count();
+           if (max_windows > 0 && window_count > max_windows){
+              throw exceptions::syntax_exception(sstring("The number of windows (" + std::to_string(window_count) + ") specified is larger than the allowed "
```
Use fmt::format() for string formatting, it's less terrible than string concatenation.

avikivity commented on db/config.cc:
```diff
@@ -840,6 +840,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , max_memory_for_unlimited_query_hard_limit(this, "max_memory_for_unlimited_query_hard_limit", "max_memory_for_unlimited_query", liveness::LiveUpdate, value_status::Used, (uint64_t(100) << 20),
             "Maximum amount of memory a query, whose memory consumption is not naturally limited, is allowed to consume, e.g. non-paged and reverse queries. "
             "This is the hard limit, queries violating this limit will be aborted.")
+    , twcs_max_window_count(this, "twcs_max_window_count", value_status::Used, 50, "The maximum number of compaction windows allowed when making use of TimeWindowCompactionStrategy. A setting of 0 effectively disables the restriction.")
```
LiveUpdate so we don't need a restart when this changes.

avikivity commented on cql3/statements/create_table_statement.cc:
```diff
@@ -465,6 +465,20 @@ std::optional<sstring> check_restricted_table_properties(
                                                  "number of windows specified by the twcs_max_window_count (" + std::to_string(max_windows) + ") parameter."));
            }
+        } else {
+              switch(qp.db().get_config().restrict_twcs_default_ttl()){
```
Coding style wants space after switch (switch is not a function).

fee-mendes commented:
Changes: 

- Corrected coding-style to follow our standards
- Removed C-style case and explicit casting as it was redundant
- Switched to fmt::format() for easier readability
- `twcs_max_window_count` is not LiveUpdate-able

avikivity commented on cql3/statements/create_table_statement.cc:
```diff
@@ -452,6 +453,20 @@ std::optional<sstring> check_restricted_table_properties(
+        int max_windows = qp.db().get_config().twcs_max_window_count();
+
+        if (ttl > 0) {
+           int window_count = ttl / _twcs_options.get_sstable_window_size().count();
```
You missed this. get_sstable_window_size() is in seconds, but this code doesn't "know" that. get_default_time_to_live() has no units, so we have to know that it's in seconds. To insulate the code against get_sstable_window_size_in_seconds, write

```c++
auto window_count = std::chrono::seconds(ttl) / _twcs_options.get_sstable_window_size();
```

Now if get_sstable_window_size() returns millseconds or nanoseconds, the code still works because std::chrono::duration auto-adjusts for units.

avikivity replied:
plus, indents should be four spaces, not three.

avikivity commented:
Please add unit tests.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2064/

fee-mendes commented:
In v3:
    - window_count now properly adjusts to a std::chrono::duration should units change
    - added unit tests
    - indentation correction

nyh commented on cql3/statements/create_table_statement.cc:
```diff
@@ -452,6 +453,20 @@ std::optional<sstring> check_restricted_table_properties(
+            auto window_count = std::chrono::seconds(ttl) / _twcs_options.get_sstable_window_size();
+            if (max_windows > 0 && window_count > max_windows) {
+                throw exceptions::syntax_exception(fmt::format("The number of windows ({}) specified is larger than the allowed "
+                                                   "number of windows specified by the twcs_max_window_count ({}) parameter.", window_count, max_windows));
```
I think the error message could be, and should be, more explicit on what the user set up wrong. Since this test is only about the default_ttl case, I think the error message should tell you more directly what was tested. Something like: **"The setting of default_time_to_live={} and compaction window={} can lead to {} windows, which is larger than the allowed number of windows specified by the twcs_max_window_count ({}) configuration)."**

nyh commented on cql3/statements/create_table_statement.cc:
```diff
@@ -452,6 +453,20 @@ std::optional<sstring> check_restricted_table_properties(
+        if (ttl > 0) {
+            auto window_count = std::chrono::seconds(ttl) / _twcs_options.get_sstable_window_size();
+            if (max_windows > 0 && window_count > max_windows) {
+                throw exceptions::syntax_exception(fmt::format("The number of windows ({}) specified is larger than the allowed "
```
In the previous code which I wrote in check_restricted_table_properties() I chose to throw configuration_exception, not syntax_exception. Because it's not really a syntax error, and the query itself might not be an error but it just doesn't fit the configured twcs_max_window_count.

nyh commented on db/config.cc:
```diff
@@ -840,6 +840,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
             "Maximum amount of memory a query, whose memory consumption is not naturally limited, is allowed to consume, e.g. non-paged and reverse queries. "
             "This is the hard limit, queries violating this limit will be aborted.")
+    , twcs_max_window_count(this, "twcs_max_window_count", liveness::LiveUpdate, value_status::Used, 50,
+            "The maximum number of compaction windows allowed when making use of TimeWindowCompactionStrategy. A setting of 0 effectively disables the restriction.")
```
Note that your code above doesn't actually enforce this in the general case - only when default TTL is enabled.

fee-mendes replied:
That's correct, because tables without a TTL defined can either have a bound (provided you use `USING TTL`) or unbound window size. Ideally, one would want to always forbid creation of tables without a TTL defined - **BUT**, this alone is already enough to break tools such as cassandra-stress. And the only way to tell cassandra-stress to create a table with a defined TTL is to provide an user profile.

That said, instead of failing everything altogether (and a bunch of dtests and units), we print a warning (in another commit) other than immediately refusing the creation of the table. For production settings, however, we should enforce it.

nyh replied:
What I meant was that the option string isn't 100% accurate because it doesn't mention that this restriction only applies to tables with a default ttl. But this is a minor quibble, let's not delay the merge for this.

nyh commented on cql3/statements/create_table_statement.cc:
```diff
@@ -465,6 +465,20 @@ std::optional<sstring> check_restricted_table_properties(
+              case db::tri_mode_restriction_t::mode::WARN:
+                  return format("TimeWindowCompactionStrategy tables without a default_time_to_live "
+                      "may potentially introduce too many windows. Ensure that insert statements specify a "
+                      "TTL (via USING TTL), when inserting data to this table");
```
I suggest adding to this string also a phrase like "The ??? configuration option can be changed to silence this warning  or make it into an error".

nyh commented on cql3/statements/create_table_statement.cc:
```diff
@@ -465,6 +465,20 @@ std::optional<sstring> check_restricted_table_properties(
                                                    "number of windows specified by the twcs_max_window_count ({}) parameter.", window_count, max_windows));
             }
+        } else {
+              switch (qp.db().get_config().restrict_twcs_default_ttl()) {
```
I think the name "restrict_twcs_default_ttl" is a bit confusing, because turning it to true doesn't restrict using default_ttl - it requires default_ttl. It restricts the ability *not* to use it. So I wonder if the name "restrict_twcs_without_default_ttl" would be more accurate?

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -91,6 +91,32 @@ SEASTAR_TEST_CASE(test_create_table_with_id_statement) {
+            e.execute_cql("UPDATE system.config SET value='true' WHERE name='restrict_twcs_default_ttl';").get();
+            assert_that_failed(e.execute_cql("CREATE TABLE tbl2 (a int, b int, PRIMARY KEY (a)) WITH "
+                          "compaction = {'class': 'TimeWindowCompactionStrategy', "
+                          "'compaction_window_size': '1', 'compaction_window_unit': 'MINUTES'};"));
```
The last line only checked failure, not success.
Please add here also a test that:
1. Creating now tbl2 *with* default_time_to_live=60 will work (the restriction will not prevent it)
2. After that, trying to ALTER TABLE tbl2 to remove the default_time_to_live will fail (if I understand correctly, the new check is performed in any ALTER TABLE, it doesn't matter if default_time_to_live or the compaction strategy is what was changed).

nyh reviewed (changes requested):
Looks very good!
I just left a few comments and requests. Thanks.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2066/

nyh commented:
@fee-mendes it turns out your patch breaks one dtest - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2066/testReport/junit/resharding_test/TestReshardingSingleNodeGating/Sanity_Tests___test_resharding_by_murmur3_gating_1_TimeWindowCompactionStrategy_15_/
which it appears (if I understand the error message correctly) refuses to accept warning messages.

This is frustrating, because it means we need to fix the dtest first. The "obvious" fix - to configure the new restrict configuration to FALSE - isn't trivial because the result would not work on current versions of Scylla because they don't support this configuration. So maybe the test should be fixed to ignore the warning message?

fee-mendes commented:
In v4:
- Suggestions were addressed
- Incorrect syntax_exception was switched to configuration_exception
- restrict_twcs_default_ttl() was renamed to restrict_twcs_without_default_ttl() for accuracy
- added more unit tests and conditions
- `ALTER TABLE` statements without specifying a TTL and/or the underlying CompactionStrategy could potentially bypass all checks, this situation is now handled and an additional unit test is made available to check 10 distinct situations

I am still stuck with the unit testing, will see if I arrange time to follow-up on it today. So it will fail on the same dtest anyway for now. :clown_face:

raphaelsc commented on cql3/statements/create_table_statement.cc:
```diff
@@ -434,6 +436,23 @@ std::optional<sstring> check_restricted_table_properties(
+        current_strategy = sstables::compaction_strategy_type::null;
+        current_ttl = gc_clock::duration::zero();
+    }
+
```
Better to add an optional schema (std::optional<schema>) in check_restricted_table_properties() parameter list. So we won't have to embed this knowledge in check_restricted_table_properties(). So alter table calls it with schema, whereas create table calls it with std::nullopt.

fee-mendes replied:
I agree. That's what happens when an Architect pretends to add things. I will change it once I get a consensus on the failing dtest thingy, there's no point on pushing changes when these are going to fail anyway on CI.

raphaelsc replied:
Don't worry. You did a great job with this series. This protection is very important for the product.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2102/

psarna commented on test/boost/cql_query_test.cc:
```diff
@@ -91,6 +91,38 @@ SEASTAR_TEST_CASE(test_create_table_with_id_statement) {
 
+SEASTAR_TEST_CASE(test_create_twcs_table_no_ttl) {
+    return do_with_cql_env([](cql_test_env& e) {
+        return seastar::async([&e] {
```
nitpick: there's also `do_with_cql_env_thread`, which lets you avoid explicitly calling a thread

fee-mendes commented:
In v5:

- check_restricted_table_properties() parameter list now expects and optional schema
- A corner-case on which one could try to update an unrelated table property in a TWCS table without TTL defined when `restrict_twcs_without_default_ttl` is set to true would make the change impossible without LiveUpdate'ing. Similarly, when `restrict_twcs_default_ttl` is set to warn, any `ALTER TABLE` to the given table would always print warnings. This situation was already addressed when checking the window_count but not when TTL didn't exist. The fix was simply to move the check one level above, early in the processing. The unit test was extended to cover this situation.
- The relevant unit tests now call `do_with_cql_env_thread` instead of `do_with_cql_env`

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2132/

raphaelsc commented on cql3/statements/create_table_statement.cc:
```diff
@@ -426,6 +427,7 @@ void create_table_statement::raw_statement::add_column_alias(::shared_ptr<column
 std::optional<sstring> check_restricted_table_properties(
     query_processor& qp,
+    std::optional<schema_ptr> schema,
     const sstring& keyspace, const sstring& table,
```
we need to move this change (which adds optional schema in function declaration) to the previous patch. otherwise, we break bisectability.

fee-mendes replied:
Right, `rebase`'ing weirdos. :roll_eyes:

fee-mendes replied:
All set, flagged as resolved.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2139/

raphaelsc commented on cql3/statements/create_table_statement.cc:
```diff
@@ -453,6 +467,27 @@ std::optional<sstring> check_restricted_table_properties(
+        sstables::time_window_compaction_strategy_options _twcs_options(options);
+        long ttl = (cfprops.has_property(cf_prop_defs::KW_DEFAULT_TIME_TO_LIVE)) ? cfprops.get_default_time_to_live() : current_ttl.count();
+        int max_windows = qp.db().get_config().twcs_max_window_count();
+
```
uint32_t or just use auto.

raphaelsc commented on cql3/statements/create_table_statement.cc:
```diff
@@ -453,6 +467,27 @@ std::optional<sstring> check_restricted_table_properties(
+        if (ttl > 0) {
+            auto window_count = std::chrono::seconds(ttl) / _twcs_options.get_sstable_window_size();
+            if (max_windows > 0 && window_count > max_windows) {
+                throw exceptions::configuration_exception(fmt::format("The setting of default_time_to_live={} and compaction window={} "
```
please append window={} with s to indicate the unit is seconds.

raphaelsc reviewed (changes requested):
Only had a few minor comments. Looks good.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2186/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2187/

fee-mendes commented:
After a 1 week saga pushing for dtests fixes to pass, this series now got through our CI. :-) 

@nyh I believe I have already covered all your suggestions and other scenarios, mind a last review and merging as needed? Thanks.

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -125,6 +125,34 @@ SEASTAR_TEST_CASE(test_create_twcs_table_no_ttl) {
+SEASTAR_TEST_CASE(test_twcs_max_window) {
+    return do_with_cql_env_thread([](cql_test_env& e) {
+        // Create a TWCS table with no defined TTL
+        e.execute_cql("CREATE TABLE tbl (a int, b int, PRIMARY KEY (a)) WITH "
```
I don't think you need this specific check here - I think you already have it in the other test in the previous patch (test_create_twcs_table_no_ttl) that deals with no TTL at all.

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -125,6 +125,34 @@ SEASTAR_TEST_CASE(test_create_twcs_table_no_ttl) {
+        e.execute_cql("CREATE TABLE tbl2 (a int, b int, PRIMARY KEY (a)) WITH "
+                      "compaction = {'class': 'TimeWindowCompactionStrategy', "
+                      "'compaction_window_size': '1', 'compaction_window_unit': 'HOURS'} "
+                      "AND default_time_to_live=86400;").get();
```
This test depends on a specific setting of the default of the restriction (50). I wonder if the test shouldn't specify this restriction explicitly - otherwise if we change our mind in the future what the default should be, this test will break.
Or, on the other hand, maybe it's good to have this test to verify that the default can't change unintentionally?

fee-mendes replied:
We now early LiveUpdate to ensure if we decide to change the default the tests won't break. ;-)

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -153,6 +153,74 @@ SEASTAR_TEST_CASE(test_twcs_max_window) {
+
+        // Cenario 2: STCS->TWCS with small TTL
+        e.execute_cql("CREATE TABLE tbl2 (a int PRIMARY KEY, b int) WITH default_time_to_live = 60;").get();
+        e.execute_cql("ALTER TABLE tbl2 WITH compaction = {'class': 'TimeWindowCompactionStrategy'};").get();
```
Hmm, do we have a default TWCS window size? I didn't realize this. What is it?

fee-mendes replied:
Yes, we do. 1 day.

```cpp
    std::chrono::seconds sstable_window_size = DEFAULT_COMPACTION_WINDOW_UNIT * DEFAULT_COMPACTION_WINDOW_SIZE;
```

and later on:
```cpp
    static constexpr std::chrono::seconds DEFAULT_COMPACTION_WINDOW_UNIT = 86400s;
    static constexpr int DEFAULT_COMPACTION_WINDOW_SIZE = 1;
```

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -153,6 +153,74 @@ SEASTAR_TEST_CASE(test_twcs_max_window) {
+        e.execute_cql("CREATE TABLE tbl4 (a int PRIMARY KEY, b int) WITH compaction = "
+                      "{'class': 'TimeWindowCompactionStrategy'} AND default_time_to_live = 60;").get();
+        e.require_table_exists("ks", "tbl4").get();
+        assert_that_failed(e.execute_cql("ALTER TABLE tbl4 WITH default_time_to_live = 86400000;"));
```
In the code patches, I was wondering if we are handling the TWCS options coming from the pre-existing table correctly. Here I think this isn't checked properly - because you have a TWCS table with the default window option, so since it's the default you can't know if the empty options or the correct options are being checked.
I think you should explicitly list a TWCS window size in some of these checks (or add cases with an explict TWCS window size)

fee-mendes replied:
By TWCS options I assume you mean the introduced restrictions? If so, yes, I agree to "hardcode" it on the test, specially should we want to change the defaults as you had mentioned a few comments earlier.

However, since your comment says  `from the pre-existing table correctly`, this got me confused. The code is quite simple: If the user is switching to TWCS and/or modifying `default_time_to_live` then we run the conditions. If the user is switching to TWCS and NOT specifying a `default_time_to_live`, then we use the already defined table TTL. For any other condition where the user is specifying the `default_time_to_live` we check against the user provided input.

The reason is quite simple: If we don't do that (ie: We always check from the pre-existing table), then we block any other table `ALTER` statement needlessly. We don't want an `ALTER TABLE x WITH comment='bla'` to fail just because the table was incorrectly created from the beginning, right? ;-)

nyh replied:
No, I meant the `sstable_window_size`.

When I reviewed the code above, I was worried (and wrote a comment) that you always take `sstable_window_size` from cfprops (the ALTER command), never from the existing table. This means (I think) that if the table was previously created with some non-default `sstable_window_size`, and now you get an ALTER command which modifies `default_time_to_live`, you calculates will use the default `sstable_window_size` instead of the actual one.

After suspecting this, I looked at your test and saw that you use the default `sstable_window_size` in these tests, so they won't be sensitive to such a bug if it's a real bug.

fee-mendes replied:
Ok, it makes sense now. You're right, a change in `default_time_to_live` is currently checked against the default, duh. I'll prepare a fix.

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -153,6 +153,74 @@ SEASTAR_TEST_CASE(test_twcs_max_window) {
+        e.execute_cql("CREATE TABLE tbl6 (a int PRIMARY KEY, b int) WITH compaction = "
+                      "{'class': 'TimeWindowCompactionStrategy'} AND default_time_to_live = 86400000;").get();
+        e.require_table_exists("ks", "tbl6").get();
+        e.execute_cql("UPDATE system.config SET value='50' WHERE name='twcs_max_window_count';").get();
```
Isn't 50 the default of this parameter?

fee-mendes replied:
Correct, I've added a smaller value at the beginning of the test so I didn't have to rewrite it all. Also - it prevents some sections from failing if we decided to change the default in the future.

nyh reviewed (changes requested):
Looks mostly good, but I left a few suggestions.
In one place I'm afraid there might be a bug (pre-existing TWCS window size setting not consulted) and I think it's not covered by the test - but maybe I just misunderstood something?

fee-mendes commented on cql3/statements/create_table_statement.cc:
```diff
@@ -453,6 +467,27 @@ std::optional<sstring> check_restricted_table_properties(
+        }
+
+        if (ttl > 0) {
+            auto window_count = std::chrono::seconds(ttl) / _twcs_options.get_sstable_window_size();
```
You're right. I was under the impression `get_sstable_window_size()` could never be 0. But for whatever reason we allow the user to explicitly set `'compaction_window_size': '0'` :clown_face: 

This should have been strictly forbidden - however - so it is good you realized it, I'll cook a fix. 

The funny thing is that what you just reported here is actually a Scylla more obscure bug. If we stop and think: What does a 0 size window does? Does it create a window for every 0 seconds or does it behaves as a "single-window" STCS table?

Cassandra properly refuses the creation of such table:

```cql
cassandra@cqlsh> CREATE TABLE ks.twcs (pk int primary key) WITH compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': 0};
ConfigurationException: 0 must be greater than 1 for compaction_window_size
```

Scylla accepts it in mysterious ways:
```cql
cassandra@cqlsh> CREATE TABLE ks.twcs (pk int primary key) WITH compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': 0, 'compaction_window_unit': 'MINUTES'};
cassandra@cqlsh>
```

:face_exhaling: @raphaelsc

fee-mendes replied:
Well, we not only accept 0's but also negative values, which makes it even more interesting. I'll fire up a separate issue for that.

raphaelsc replied:
> Well, we not only accept 0's but also negative values, which makes it even more interesting. I'll fire up a separate issue for that.
> 
> 

No need to open an issue for that. Lack of validation on compaction strategy params is a known issue and we have an issue opened tracking it already

fee-mendes replied:
https://github.com/scylladb/scylladb/issues/11521

fee-mendes replied:
The division by zero potential situation is now handled. I also added a commit to fix it for good as it is really troubling me this has been lying around for so long.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2191/

fee-mendes commented:
In v6:
- Added a commit to reject table creations with an invalid compaction_window_size. This - in turn - aligns the behavior to the same as Cassandra. It may be decided to drop this specific commit from this series, although I'd argue that it fits in as we are discussing guardrails and - thus - provide options to avoid users from shooting themselves in their feet.
- A potential division by zero situation was corrected. Unfortunately, since we have long been exposed by the fact users could create tables with an invalid window_size we can not simply ignore that fact. Due to that, when a window_size <= 0 is found, we'll default it to 1 minute, which is TWCS smaller window size possible.
- The series now properly references the TWCS table properties for already created tables, previously it only would use the command-line provided options, which was wrong - given that ALTER TABLE statements providing only a default_time_to_live would retrieve TWCS default settings
- Unit tests now specify a specific value for `twcs_max_window_count`. This avoids breaking the tests should we want to modify the default in the future. Unit tests now cover both aforementioned situations (potential division by zero and ensure we reference the already created Table properties) and some redundant tests have been removed.
- Minor var renames (_twcs_options -> twcs_options specifically)

nyh commented on test/boost/cql_query_test.cc:
```diff
@@ -150,6 +150,98 @@ SEASTAR_TEST_CASE(test_twcs_max_window) {
+                     "'compaction_window_size': '2', 'compaction_window_unit': 'MINUTES'};").get();
+       e.execute_cql("ALTER TABLE tbl11 WITH default_time_to_live=3600;").get();
+
+       // Cenario 12: Ensure that window sizes <= 0 are forbidden
```
Nitpick: you added 12 scenarios not 10 as the commit message says, and mis-spelled "scenario" in all of them ;-)
